File: cerebral-api/app/Models/Material.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;

class Material extends Model
{
    use HasFactory;

    protected $fillable = [
        'name',
        'description',
        'category',
        'unit',
        'current_stock',
        'min_stock',
        'max_stock',
        'unit_price',
        'supplier',
        'supplier_contact',
        'location',
        'status',
        'created_by',
    ];

    protected $casts = [
        'current_stock' => 'decimal:2',
        'min_stock' => 'decimal:2',
        'max_stock' => 'decimal:2',
        'unit_price' => 'decimal:2',
    ];

    /**
     * Get the user who created the material.
     */
    public function createdBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    /**
     * Get the deliveries associated with this material.
     */
    public function deliveries(): BelongsToMany
    {
        return $this->belongsToMany(Delivery::class, 'delivery_material')
            ->withPivot('quantity', 'received_quantity', 'notes')
            ->withTimestamps();
    }

    /**
     * Check if the material is in low stock.
     */
    public function isLowStock(): bool
    {
        return $this->current_stock <= $this->min_stock;
    }

    /**
     * Check if the material is out of stock.
     */
    public function isOutOfStock(): bool
    {
        return $this->current_stock <= 0;
    }

    /**
     * Update the stock level.
     */
    public function updateStock(float $quantity, string $operation = 'add'): void
    {
        if ($operation === 'add') {
            $this->current_stock += $quantity;
        } elseif ($operation === 'subtract') {
            $this->current_stock = max(0, $this->current_stock - $quantity);
        }

        $this->updateStatus();
        $this->save();
    }

    /**
     * Update the status based on current stock.
     */
    protected function updateStatus(): void
    {
        if ($this->current_stock <= 0) {
            $this->status = 'out_of_stock';
        } elseif ($this->isLowStock()) {
            $this->status = 'low_stock';
        } else {
            $this->status = 'available';
        }
    }

    /**
     * Get the stock percentage.
     */
    public function getStockPercentageAttribute(): float
    {
        if ($this->max_stock <= 0) {
            return 0;
        }

        return ($this->current_stock / $this->max_stock) * 100;
    }
}

File: cerebral-api/lib/models/material.dart
class Material {
  final int id;
  final String name;
  final String description;
  final String category;
  final String unit;
  final int currentStock;
  final int minimumStock;
  final int maximumStock;
  final double unitPrice;
  final String? supplier;
  final String? location;
  final String status;
  final DateTime? lastUpdated;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Material({
    required this.id,
    required this.name,
    required this.description,
    required this.category,
    required this.unit,
    required this.currentStock,
    required this.minimumStock,
    required this.maximumStock,
    required this.unitPrice,
    this.supplier,
    this.location,
    required this.status,
    this.lastUpdated,
    this.createdAt,
    this.updatedAt,
  });

  factory Material.fromJson(Map<String, dynamic> json) {
    return Material(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      category: json['category'],
      unit: json['unit'],
      currentStock: json['current_stock'],
      minimumStock: json['minimum_stock'],
      maximumStock: json['maximum_stock'],
      unitPrice: double.parse(json['unit_price'].toString()),
      supplier: json['supplier'],
      location: json['location'],
      status: json['status'],
      lastUpdated: json['last_updated'] != null ? DateTime.parse(json['last_updated']) : null,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'category': category,
      'unit': unit,
      'current_stock': currentStock,
      'minimum_stock': minimumStock,
      'maximum_stock': maximumStock,
      'unit_price': unitPrice,
      'supplier': supplier,
      'location': location,
      'status': status,
      'last_updated': lastUpdated?.toIso8601String(),
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  bool get isLowStock => currentStock <= minimumStock;
  bool get isOutOfStock => currentStock == 0;
  bool get isOverstocked => currentStock > maximumStock;
  bool get isAvailable => currentStock > 0;
  
  double get stockPercentage => maximumStock > 0 ? (currentStock / maximumStock) * 100 : 0;
  double get totalValue => currentStock * unitPrice;
  
  String get stockStatus {
    if (isOutOfStock) return 'Rupture';
    if (isLowStock) return 'Stock faible';
    if (isOverstocked) return 'Surstock';
    return 'Normal';
  }
  
  String get stockStatusColor {
    if (isOutOfStock) return '#F44336';
    if (isLowStock) return '#FF9800';
    if (isOverstocked) return '#9C27B0';
    return '#4CAF50';
  }
}
